## TP : Règles d'associations


Nom et prénom : [**BEYER-BONNIARD Lucas**](https://github.com/Ptitlu42)



(a) Lire le fichier **"ratings_clean_names.csv"**

(b) Transformez le tableau pour obtenir une structure où les utilisateurs correspondent aux "transactions" (lignes) et les films aux articles (colonnes).  

(c) Convertissez le tableau en un ensemble de données binaire où 0 signifie "non visionné" et 1 signifie "visionné".  

(d) Utilisez la bibliothèque **mlxtend** pour extraire les motifs fréquents.   
Installer la version 0.23.1  **pip install mlxtend==0.23.1**
Découvrir les ensembles d'éléments fréquents. Par exemple, utilisez les paramètres suivants :  
   - `min_support=0.05`,  
   - `max_len=3`,  
   - et l'option `use_colnames=True` pour des résultats facilement compréhensibles.  
   (Le calcul prend moins d'une minute).  

(e) Affichez les 10 ensembles d'éléments les plus fréquents.  
   Vous pouvez utiliser `pd.set_option('display.max_colwidth', None)` pour forcer l'affichage complet des colonnes.  

(f) Comptez la taille des ensembles d'éléments.  

(g) Filtrer pour afficher les ensembles de 2 éléments les plus fréquents, puis  les ensembles de 1 élément les plus fréquents.

(h) Qu'est ce que vous remarquez ? 

## Règles d'association

(a) Utilisez `mlxtend.frequent_patterns.association_rules` pour calculer les scores typiques des motifs fréquents.  

(b) Triez les règles d'association par leur **lift** et interprétez les observations.  

(c) Recherchez les antécédents qui expliquent le mieux un conséquent donné.

(d) Explorez les résultats en ajustant différents seuils pour le support, la longueur maximale et les scores des règles d'association afin de trouver des explications pertinentes sur les raisons pour lesquelles certaines personnes qui aiment certains films ont tendance à en aimer d'autres.  


# TP


### (a) Lire le fichier **"ratings_clean_names.csv"**



In [6]:
%pip install mlxtend
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

/home/ptitlu/Data/code/cours/fouille_donnée/.conda/lib/python3.12/pty.py:95: DeprecationWarning: This process (pid=20940) is multi-threaded, use of forkpty() may lead to deadlocks in the child.
  pid, fd = os.forkpty()


Note: you may need to restart the kernel to use updated packages.


In [7]:
df = pd.read_csv('ratings_clean_names.csv')
df.head()

,userId,rating,movieId,timestamp,title_safe
0,2038,2.0,18,1305046590,Four Rooms(5.0)
1,2858,4.0,18,856639289,Four Rooms(5.0)
2,2928,4.0,18,862900088,Four Rooms(5.0)
3,5687,4.0,18,865421714,Four Rooms(5.0)
4,6795,2.5,18,1336325802,Four Rooms(5.0)


### (b) Transformez le tableau pour obtenir une structure où les utilisateurs correspondent aux "transactions" (lignes) et les films aux articles (colonnes).  


In [10]:
pivot_table = df.pivot_table(
    index='userId',    
    columns='title_safe',   
    values='rating',   
    aggfunc='mean',    
    fill_value=0       
)

print("Dimensions de la table pivot :", pivot_table.shape)
print("\nAperçu des premières lignes :")
pivot_table

Dimensions de la table pivot : (17906, 987)

Aperçu des premières lignes :


title_safe,(500) Days of Summer(19913.0),10 Things I Hate About You(4951.0),101 Dalmatians(11674.0),12 Angry Men(389.0),2 Days in the Valley(9401.0),"20,000 Leagues Under the Sea(173.0)",2001: A Space Odyssey(62.0),28 Days Later(170.0),300(1271.0),A Beautiful Mind(453.0),...,Wyatt Earp(12160.0),X-Men(36657.0),X-Men: Days of Future Past(127585.0),X-Men: First Class(49538.0),X2(36658.0),You've Got Mail(9489.0),Young Frankenstein(3034.0),Young Guns(11967.0),Zombieland(19908.0),Zootopia(269149.0)
userId,,,,,,,,,,,,,,,,,,,,,
22,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
25,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
33,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
35,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
77,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270830,0.0,0.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,4.0,...,0.0,3.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
270832,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
270834,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


### (c) Convertissez le tableau en un ensemble de données binaire où 0 signifie "non visionné" et 1 signifie "visionné".  


In [12]:
# Conversion en données binaires (0 = non visionné, 1 = visionné)
binary_data = (pivot_table > 0).astype(int)

# Affichage des premières lignes pour vérifier le résultat
print("Dimensions des données binaires :", binary_data.shape)
print("\nAperçu des données binaires :")
binary_data

Dimensions des données binaires : (17906, 987)

Aperçu des données binaires :


title_safe,(500) Days of Summer(19913.0),10 Things I Hate About You(4951.0),101 Dalmatians(11674.0),12 Angry Men(389.0),2 Days in the Valley(9401.0),"20,000 Leagues Under the Sea(173.0)",2001: A Space Odyssey(62.0),28 Days Later(170.0),300(1271.0),A Beautiful Mind(453.0),...,Wyatt Earp(12160.0),X-Men(36657.0),X-Men: Days of Future Past(127585.0),X-Men: First Class(49538.0),X2(36658.0),You've Got Mail(9489.0),Young Frankenstein(3034.0),Young Guns(11967.0),Zombieland(19908.0),Zootopia(269149.0)
userId,,,,,,,,,,,,,,,,,,,,,
22,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
25,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
33,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
35,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
77,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270830,0,0,1,0,0,0,0,0,0,1,...,0,1,0,0,0,0,0,0,0,0
270832,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
270834,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


### (d) Utilisez la bibliothèque **mlxtend** pour extraire les motifs fréquents.   



In [13]:
frequent_itemsets = apriori(binary_data, 
                           min_support=0.05,  
                           max_len=3,         
                           use_colnames=True) 

print("Nombre total d'ensembles fréquents trouvés :", len(frequent_itemsets))
print("\nAperçu des premiers ensembles fréquents :")
print(frequent_itemsets.head())

/home/ptitlu/Data/code/cours/fouille_donnée/.conda/lib/python3.12/site-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Nombre total d'ensembles fréquents trouvés : 13444

Aperçu des premiers ensembles fréquents :
    support                                  itemsets
0  0.053111                 (A Beautiful Mind(453.0))
1  0.075729                  (A Time to Kill(1645.0))
2  0.254942      (Ace Ventura: Pet Detective(3049.0))
3  0.113202  (Ace Ventura: When Nature Calls(9273.0))
4  0.145314            (Addams Family Values(2758.0))


### (e) Affichez les 10 ensembles d'éléments les plus fréquents.  


In [14]:
pd.set_option('display.max_colwidth', None)

print("Les 10 ensembles d'éléments les plus fréquents :")
print(frequent_itemsets.nlargest(10, 'support'))

Les 10 ensembles d'éléments les plus fréquents :
      support                           itemsets
102  0.448844              (Pulp Fiction(680.0))
153  0.444600  (The Shawshank Redemption(278.0))
53   0.438121               (Forrest Gump(13.0))
154  0.400648  (The Silence of the Lambs(274.0))
76   0.348989             (Jurassic Park(329.0))
11   0.336926                 (Apollo 13(568.0))
133  0.328326             (The Fugitive(5503.0))
33   0.325087        (Dances with Wolves(581.0))
16   0.324472                    (Batman(268.0))
20   0.319725                (Braveheart(197.0))


### (f) Comptez la taille des ensembles d'éléments.  


In [15]:
tailles_ensembles = frequent_itemsets.itemsets.apply(len)
comptage_tailles = tailles_ensembles.value_counts().sort_index()

print("Nombre d'ensembles par taille :")
for taille, nombre in comptage_tailles.items():
    print(f"Ensembles de taille {taille}: {nombre}")

Nombre d'ensembles par taille :
Ensembles de taille 1: 175
Ensembles de taille 2: 1763
Ensembles de taille 3: 11506


### (g) Filtrer pour afficher les ensembles de 2 éléments les plus fréquents, puis  les ensembles de 1 élément les plus fréquents.


In [16]:
ensembles_2 = frequent_itemsets[frequent_itemsets.itemsets.apply(len) == 2]
print("Les 10 ensembles de 2 éléments les plus fréquents :")
print(ensembles_2.nlargest(10, 'support'))

print("\n" + "="*80 + "\n")

ensembles_1 = frequent_itemsets[frequent_itemsets.itemsets.apply(len) == 1]
print("Les 10 ensembles de 1 élément les plus fréquents :")
print(ensembles_1.nlargest(10, 'support'))

Les 10 ensembles de 2 éléments les plus fréquents :
       support  \
1580  0.279739   
1134  0.274210   
1581  0.269910   
1141  0.261532   
1170  0.253323   
1902  0.250754   
544   0.248688   
1171  0.248632   
393   0.245002   
382   0.244611   

                                                                itemsets  
1580              (The Shawshank Redemption(278.0), Pulp Fiction(680.0))  
1134                          (Jurassic Park(329.0), Forrest Gump(13.0))  
1581              (The Silence of the Lambs(274.0), Pulp Fiction(680.0))  
1141                           (Pulp Fiction(680.0), Forrest Gump(13.0))  
1170               (The Shawshank Redemption(278.0), Forrest Gump(13.0))  
1902  (The Silence of the Lambs(274.0), The Shawshank Redemption(278.0))  
544                           (Dances with Wolves(581.0), Batman(268.0))  
1171               (The Silence of the Lambs(274.0), Forrest Gump(13.0))  
393                        (Dances with Wolves(581.0), Apollo 13(568.0))  

### (h) Qu'est ce que vous remarquez ? 


#### Sur la structure des données

- Nous avons 17906 utilisateurs et 987 films différents
- La matrice est assez creuse (beaucoup de 0), ce qui est normal car chaque utilisateur n'a pas vu tous les films

#### Sur les ensembles fréquents

- Nous avons trouvé 13444 ensembles fréquents au total
- La répartition des tailles est intéressante :
  - 175 ensembles de taille 1 (films individuels)
  - 1763 ensembles de taille 2 (paires de films) 
  - 11506 ensembles de taille 3 (triplets de films)

#### Sur les films les plus populaires (ensembles de taille 1)

- "Pulp Fiction" est le film le plus vu (44.8% des utilisateurs)
- "The Shawshank Redemption" suit de près (44.4%)
- "Forrest Gump" complète le podium (43.8%)
- Ces films sont tous des classiques très reconnus des années 90

#### Sur les paires de films les plus fréquentes

- La paire la plus fréquente est "The Shawshank Redemption" et "Pulp Fiction" (27.9%)
- Les combinaisons impliquent souvent les mêmes films populaires
- On remarque une forte corrélation entre les films de même genre ou de même époque

#### Observations générales

- Les films les plus populaires sont majoritairement des films des années 90
- Il y a une forte concentration sur certains films "cultes"


# Règles d'association


## (a) Utilisez `mlxtend.frequent_patterns.association_rules` pour calculer les scores typiques des motifs fréquents.  


In [17]:
rules = association_rules(frequent_itemsets,
                         metric="confidence",
                         min_threshold=0.5)

print("Nombre total de règles trouvées :", len(rules))
print("\nAperçu des règles d'association (premières lignes) :")
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head())

print("\nStatistiques descriptives des métriques :")
print(rules[['support', 'confidence', 'lift']].describe())

Nombre total de règles trouvées : 33286

Aperçu des règles d'association (premières lignes) :
                antecedents                   consequents   support  \
0  (A Time to Kill(1645.0))     (Independence Day(602.0))  0.061209   
1  (A Time to Kill(1645.0))  (Mission: Impossible(954.0))  0.054228   
2  (A Time to Kill(1645.0))             (Star Wars(11.0))  0.050598   
3  (A Time to Kill(1645.0))            (The Rock(9802.0))  0.053278   
4  (A Time to Kill(1645.0))              (Twister(664.0))  0.056629   

   confidence      lift  
0    0.808260  3.414983  
1    0.716077  4.126833  
2    0.668142  2.208555  
3    0.703540  4.762792  
4    0.747788  4.304045  

Statistiques descriptives des métriques :
            support    confidence          lift
count  33286.000000  33286.000000  33286.000000
mean       0.085511      0.672725      2.775188
std        0.028482      0.109867      0.739860
min        0.050039      0.500000      1.126472
25%        0.062828      0.582563      2

## (b) Triez les règles d'association par leur **lift** et interprétez les observations.  


In [18]:
rules_sorted_by_lift = rules.sort_values('lift', ascending=False)

print("Les 10 règles d'association avec le plus grand lift :")
print(rules_sorted_by_lift[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head(10))

print("\nStatistiques sur les règles avec un lift > 2 :")
rules_high_lift = rules[rules['lift'] > 2]
print(f"Nombre de règles avec lift > 2 : {len(rules_high_lift)}")
print("\nMoyenne des métriques pour ces règles :")
print(rules_high_lift[['support', 'confidence', 'lift']].mean())

Les 10 règles d'association avec le plus grand lift :
                                                    antecedents  \
18222                              (Executive Decision(2320.0))   
18221                    (Eraser(9268.0), Broken Arrow(9208.0))   
25540                                          (Eraser(9268.0))   
25536            (Executive Decision(2320.0), The Rock(9802.0))   
25531  (Executive Decision(2320.0), Mission: Impossible(954.0))   
25535                                          (Eraser(9268.0))   
25538                        (Eraser(9268.0), The Rock(9802.0))   
25539                              (Executive Decision(2320.0))   
25543                          (Eraser(9268.0), Twister(664.0))   
25544                              (Executive Decision(2320.0))   

                                                    consequents   support  \
18222                    (Eraser(9268.0), Broken Arrow(9208.0))  0.050542   
18221                              (Executive Decision

## (c) Recherchez les antécédents qui expliquent le mieux un conséquent donné.


In [21]:
def analyze_rules_for_movie(rules_df, target_movie):
    filtered_rules = rules_df[rules_df['consequents'].apply(lambda x: target_movie in x)]
    
    filtered_rules = filtered_rules.sort_values('lift', ascending=False)
    
    print(f"\nMeilleures règles pour prédire le visionnage de {target_movie}:")
    print("\nTop 5 des règles par lift:")
    print(filtered_rules[['antecedents', 'confidence', 'lift']].head())
    
    return filtered_rules

target_movie = "Pulp Fiction(680.0)"

pulp_fiction_rules = analyze_rules_for_movie(rules, target_movie)

print("\nStatistiques détaillées pour les règles:")
print("Nombre de règles trouvées:", len(pulp_fiction_rules))
print("\nRègles avec la plus haute confiance:")
print(pulp_fiction_rules.nlargest(5, 'confidence')[['antecedents', 'confidence', 'lift']])
print("\nRègles avec le plus haut support:")
print(pulp_fiction_rules.nlargest(5, 'support')[['antecedents', 'support', 'confidence', 'lift']])


Meilleures règles pour prédire le visionnage de Pulp Fiction(680.0):

Top 5 des règles par lift:
                               antecedents  confidence      lift
27253                       (Ghost(251.0))    0.512639  4.851650
28480        (Natural Born Killers(241.0))    0.587527  3.980420
28479  (Interview with the Vampire(628.0))    0.507189  3.718969
20510        (Natural Born Killers(241.0))    0.562013  3.688930
30213        (Natural Born Killers(241.0))    0.518072  3.643599

Statistiques détaillées pour les règles:
Nombre de règles trouvées: 1447

Règles avec la plus haute confiance:
                                                          antecedents  \
30220        (The Usual Suspects(629.0), Natural Born Killers(241.0))   
28620  (Interview with the Vampire(628.0), The Usual Suspects(629.0))   
30187                     (Se7en(807.0), Natural Born Killers(241.0))   
27003               (Get Shorty(8012.0), Natural Born Killers(241.0))   
30186               (Quiz Show(1145